# Genomic evidence of contemporary hybridization between Schistosoma species

This repository contains the data and the analysis workflow for the manuscript "Genomic evidence of contemporary hybridization between Schistosoma species".

The repository is divided into 3 sections:
1). Production - containing all the commands used to produce data for analyses.
2). Analysis - containing all the commands used to analyse data.
3). Figures - containing scripts (1 script per figure) used to plot the results.

## Abstract
Hybridization between different species of parasites is increasingly being recognised as a public and veterinary health risk with the potential to impact the virulence and epidemiology of parasitic infections. Recent research has revealed that viable hybrids and introgressed lineages between Schistosoma spp are prevalent across Africa and beyond, including those with zoonotic potential. However, it is unclear whether these hybrid lineages represent recent hybridization events, suggesting the hybridization is frequent and ongoing, or whether these represent introgressed lineages derived from ancient hybridization events. To characterize the composition of naturally occurring schistosome hybrids we assembled a chromosomal-scale draft assembly of Schistosoma curassoni and performed whole-genome sequencing of 21 natural livestock infective schistosome isolates. Genomic analyses identified isolates of S. bovis, S. curassoni and hybrids between the two species all of which were early generation hybrids (F1, F2, F1 backcrosses and F2 backcrosses) with multiple generations found within the same host. These results show that hybridization is an ongoing process within natural populations with the potential to further challenge elimination efforts against schistosomiasis. 
